### PR TITLE
Websocket-chat

### PR DESCRIPTION
**Pull request by sdelamo**
I move this demo from moved from micronaut-examples. Can you transform it into a guide?

Tasks: 

- [x] add metadata.json
- [x] add asciidoc file
- [x] write guide, add callouts to files and explain
- [ ] add kotlin variation
- [x] add java variation
- [ ] add groovy variation

You will need to write the java test with Junit. Currently, it is writtten with spock. 

it should use the standard package name we use in every other guide. 

We write tests for the Java version with Junit, for Groovy verison with Spock and with Kotlin version with Junit.

### CODE REVIEW

**Review by burtbeckwith — APPROVED** [review]

**Comment by sdelamo** [comment]
@hollingsworthd I have moved the test to use [Awailitility](https://github.com/awaitility/awaitility) so that it is less brittle. 

Moreover, I have moved any test related logic to the test classpath.

I have split different changes in different commits, please check them out and let me know if you have any concerns.

**Comment by sdelamo** [comment]
@burtbeckwith since you reviewed, I have added a graalvm section and made the test less brittle. Can you review it again? Thanks.

**Review by burtbeckwith — APPROVED** [review]

**Review by hollingsworthd — DISMISSED** [review]

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 1]
```diff
@@ -0,0 +1,88 @@
+common:header.adoc[]
+
+The https://datatracker.ietf.org/doc/html/rfc6455[WebSocket Protocol] allows for web browsers to establish interactive sessions with a server that are event-driven. This technology is ideal for applications that need state changes without the overhead and latency of polling the server. The article https://www.linode.com/docs/guides/introduction-to-websockets/["Introduction to WebSockets"] explains the benefits of WebSocket in more depth along with a discussion of the client-side WebSocket API.
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
The https://datatracker.ietf.org/doc/html/rfc6455[WebSocket Protocol] allows for web browsers to establish interactive sessions with a server that are event-driven. This technology is ideal for applications that need state changes without the overhead and latency of polling the server. The article https://www.linode.com/docs/guides/introduction-to-websockets/["Introduction to WebSockets"] explains the benefits of WebSocket in more depth, along with a discussion of the client-side WebSocket API.
```

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+=== Chat server
+
+Our chat server is very simple. It merely allows you to connect and broadcast messages to subscribers of the topic. There's also a special topic called "all" that can make announcements and receive messages from all topics.
+
+source:ChatServerWebSocket[]
+
+callout:websocket-server[number=1]
+callout:constructor-di[number=2,arg0=ChatServerWebSocket]
+callout:websocket-on-open[number=3]
+<4> Our chat server will have a special topic called "all" that can make announcements and receive messages from all topics
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
<4> Our chat server has a special topic called "all" that can make announcements and receive messages from all topics
```

This repeats information in line 34. Not sure if that's intentional.

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+callout:websocket-on-message[number=5]
+callout:websocket-on-close[number=6]
+<7> A predicate to keep messages within topics yet also enable the special "all" topic
+
+=== Test
+
+common:awaitility.adoc[]
+
+Micronaut Frameworks ease the creation of Websocket servers and clients. Write a test which uses `ClientWebSocket` to test the application.
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
The Micronaut frameworks eases the creation of Websocket servers and clients. 

Write a test that uses `ClientWebSocket` to test the application.
```

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+test:ChatServerWebSocketTest[]
+
+callout:spec-name[1]
+callout:micronaut-test[2]
+callout:websocket-client[3]
+callout:websocket-client-autocloseable[4]
+callout:websocket-on-message[5]
+<6> Any method name that start with (or equals) "broadcast" or "send" will be used to broadcast messages as long as the parameter is a String or a bean
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
<6> Any method name that starts with (or equals) "broadcast" or "send" will be used to broadcast messages, as long as the parameter is a String or a bean
```

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+callout:websocket-client-connect[7]
+<8> Set up WebSocket clients for multiple users
+callout:awaitility[7]
+<10> Test that the correct messages are automatically sent when users join topics
+<11> Test that chats are sent within a topic and not ouside of it except to the special "all" topic
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
<11> Test that chats are sent within a topic and not outside of it, with the exception of those sent using the special "all" topic
```

File: guides/micronaut-websocket/micronaut-websocket.adoc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+<12> Test that the correct messages are automatically sent when users leave topics
+
+
+common:testApp.adoc[]
+
+common:runapp.adoc[]
+
+Open a browser and visit a URL such as http://localhost:8080/&num;/java/Joe[http://localhost:8080/\#/java/Joe] and also in another browser tab open http://localhost:8080/&num;/java/Moka[http://localhost:8080/\#/java/Moka]. You can then try sending chats as Joe and Moka.
```
Comment (kroegerk) [inline guides/micronaut-websocket/micronaut-websocket.adoc]:
```suggestion
Open a browser and visit a URL such as http://localhost:8080/&num;/java/Joe[http://localhost:8080/\#/java/Joe]. Then in another browser tab, open http://localhost:8080/&num;/java/Moka[http://localhost:8080/\#/java/Moka]. You can then try sending chats as Joe and Moka.
```

File: guides/micronaut-websocket/src/main/resources/public/style.css [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+html {
+    overflow-y: scroll;
+}
+
+body {
+    font-family: monospace;
+    font-size: 14px;
+    max-width: 480px;
+    margin: 0 auto;
+    padding: 20px
```
Comment (kroegerk) [inline guides/micronaut-websocket/src/main/resources/public/style.css]:
```suggestion
    padding: 20px;
```

File: guides/micronaut-websocket/src/main/resources/public/style.css [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    float: right;
+}
+
+li {
+    margin: 5px 0;
+}
+
+#chatControls {
+    overflow: auto;
+    margin: 0 0 5px 0
```
Comment (kroegerk) [inline guides/micronaut-websocket/src/main/resources/public/style.css]:
```suggestion
    margin: 0 0 5px 0;
```

File: guides/micronaut-websocket/src/main/resources/public/style.css [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    margin-left: 250px;
+    background: #f0f0f9;
+    padding: 5px 10px;
+    width: 150px;
+    top: 11px;
+}
+
+#chat p {
+    margin: 5px 0;
+    font-weight: 300
```
Comment (kroegerk) [inline guides/micronaut-websocket/src/main/resources/public/style.css]:
```suggestion
    font-weight: 300;
```

File: src/docs/common/callouts/callout-awaitility.adoc [hunk 10]
```diff
@@ -0,0 +1 @@
+Use https://github.com/awaitility/awaitility[Awaitility] to test asynchronous code. By default, `Awaitility` waits for 10 seconds and if the condition is not fulfilled during this time it'll throw a `ConditionTimeoutException` failing the test.
```
Comment (kroegerk) [inline src/docs/common/callouts/callout-awaitility.adoc]:
```suggestion
Use https://github.com/awaitility/awaitility[Awaitility] to test asynchronous code. By default, `Awaitility` waits for 10 seconds and if the condition is not fulfilled during this time, it throws a `ConditionTimeoutException` failing the test.
```

File: src/docs/common/callouts/callout-static-resources.adoc [hunk 11]
```diff
@@ -0,0 +1 @@
+Configure the Framework to look for static resources in `src/main/resources/public`
```
Comment (kroegerk) [inline src/docs/common/callouts/callout-static-resources.adoc]:
```suggestion
Configure the Framework to look for static resources in `src/main/resources/public`.
```

File: src/docs/common/callouts/callout-websocket-client-autocloseable.adoc [hunk 12]
```diff
@@ -0,0 +1 @@
+The client must implement `AutoCloseable` and you should ensure that the connection is closed at some point.
```
Comment (kroegerk) [inline src/docs/common/callouts/callout-websocket-client-autocloseable.adoc]:
```suggestion
The client must implement `AutoCloseable`, and you should ensure that the connection is closed at some point.
```

File: src/docs/common/callouts/callout-websocket-client-connect.adoc [hunk 13]
```diff
@@ -0,0 +1,2 @@
+Once you have defined a client class you can connect to the client socket and start sending messages.
```
Comment (kroegerk) [inline src/docs/common/callouts/callout-websocket-client-connect.adoc]:
```suggestion
Once you have defined a client class, you can connect to the client socket and start sending messages.
```

File: guides/micronaut-websocket/java/src/test/java/example/micronaut/ChatServerWebSocketTest.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+@MicronautTest // <2>
+class ChatServerWebSocketTest {
+
+    @Inject
+    BeanContext beanContext;
+
+    @Inject
+    EmbeddedServer embeddedServer;
+
+    @Requires(property = "spec.name", value = "ChatWebSocketTest") // <1>
```
Comment (hollingsworthd) [inline guides/micronaut-websocket/java/src/test/java/example/micronaut/ChatServerWebSocketTest.java:42]: This footnote is mis-numbered

**Review by hollingsworthd — APPROVED** [review]

File: src/docs/common/snippets/common-awaitility.adoc [hunk 15]
```diff
@@ -0,0 +1,11 @@
+To test asynchronous code we use https://github.com/awaitility/awaitility[Awaitility]:
```
Comment (kroegerk) [inline src/docs/common/snippets/common-awaitility.adoc]:
```suggestion
To test asynchronous code, use https://github.com/awaitility/awaitility[Awaitility]:
```

**Pull request merged**